src/workflow_tasks.rs

```diff
@@ -492,9 +507,11 @@ impl WorkflowTaskManager {
                 }
             }
 
-            // Blow them up if we're in non-sticky mode as well
-            if self.eviction_policy == WorkflowCachingPolicy::NonSticky {
-                self.evict_run(run_id);
+            // Evict run id if we are in non-sticky mode or if we have too many cached run ids.
+            let maybe_evicted = self.cache_manager.write().insert(run_id);
+            if let Some(evicted_run_id) = maybe_evicted {
+                // Not removing from the cache as record has already been removed and we are still holding cache manager lock.
+                self.evict_run(&evicted_run_id);
```

Comment:
We've lost the nonsticky eviction condition. Also the bit in the comment about still holding the lock is irrelevant if insertion removes the one we're about to evict because we're at the limit anyway.

Reply:
It should actually work because we'll initialize cache with 0 capacity, but just in case if we've missed to touch in some situation, let's add a safety condition just in case.